File: 39/src/utils/quantumSimulator.ts
import type { Complex, GateType, QuantumCircuit, QuantumGate, SimulationResult } from '@/types/quantum';
import { GATE_MATRICES, applyMatrixToVector, createIdentityMatrix } from './quantumGates';
import { complexMagnitudeSq, complexAdd, complexMulScalar, ONE, ZERO, complexMul } from './complex';

export class QuantumSimulator {
  private qubitCount: number;
  private stateVector: Complex[];
  private dimension: number;

  constructor(qubitCount: number) {
    if (qubitCount < 1 || qubitCount > 12) {
      throw new Error('量子比特数必须在 1 到 12 之间');
    }
    this.qubitCount = qubitCount;
    this.dimension = 1 << qubitCount;
    this.stateVector = this.createInitialState();
  }

  private createInitialState(): Complex[] {
    const state: Complex[] = Array(this.dimension)
      .fill(null)
      .map(() => ({ real: 0, imag: 0 }));
    state[0] = { real: 1, imag: 0 };
    return state;
  }

  public reset(): void {
    this.stateVector = this.createInitialState();
  }

  public getStateVector(): Complex[] {
    return [...this.stateVector];
  }

  public getQubitCount(): number {
    return this.qubitCount;
  }

  public getProbabilities(): number[] {
    return this.stateVector.map((c) => complexMagnitudeSq(c));
  }

  public applySingleQubitGate(gateType: Exclude<GateType, 'CNOT'>, targetQubit: number): void {
    if (targetQubit < 0 || targetQubit >= this.qubitCount) {
      throw new Error(`无效的目标量子比特: ${targetQubit}`);
    }

    const gateMatrix = GATE_MATRICES[gateType];
    const newState: Complex[] = Array(this.dimension)
      .fill(null)
      .map(() => ({ real: 0, imag: 0 }));

    const targetMask = 1 << (this.qubitCount - 1 - targetQubit);

    for (let i = 0; i < this.dimension; i++) {
      const targetBit = (i & targetMask) !== 0 ? 1 : 0;
      const otherBits = i & ~targetMask;

      for (let row = 0; row < 2; row++) {
        const targetIndex = otherBits | (row << (this.qubitCount - 1 - targetQubit));
        const element = gateMatrix[row][targetBit];
        const amplitude = this.stateVector[i];

        newState[targetIndex] = complexAdd(newState[targetIndex], {
          real: element.real * amplitude.real - element.imag * amplitude.imag,
          imag: element.real * amplitude.imag + element.imag * amplitude.real,
        });
      }
    }

    this.stateVector = newState;
  }

  public applyCNOT(controlQubit: number, targetQubit: number): void {
    if (controlQubit === targetQubit) {
      throw new Error('控制比特和目标比特不能相同');
    }
    if (controlQubit < 0 || controlQubit >= this.qubitCount) {
      throw new Error(`无效的控制量子比特: ${controlQubit}`);
    }
    if (targetQubit < 0 || targetQubit >= this.qubitCount) {
      throw new Error(`无效的目标量子比特: ${targetQubit}`);
    }

    const newState: Complex[] = [...this.stateVector.map((c) => ({ ...c }))];

    const controlMask = 1 << (this.qubitCount - 1 - controlQubit);
    const targetMask = 1 << (this.qubitCount - 1 - targetQubit);

    for (let i = 0; i < this.dimension; i++) {
      if ((i & controlMask) !== 0) {
        const flippedIndex = i ^ targetMask;
        if (i < flippedIndex) {
          const temp = newState[i];
          newState[i] = newState[flippedIndex];
          newState[flippedIndex] = temp;
        }
      }
    }

    this.stateVector = newState;
  }

  public applyGate(gate: QuantumGate): void {
    if (gate.type === 'CNOT') {
      if (gate.controlQubit === undefined) {
        throw new Error('CNOT 门需要控制比特');
      }
      this.applyCNOT(gate.controlQubit, gate.targetQubit);
    } else {
      this.applySingleQubitGate(gate.type, gate.targetQubit);
    }
  }

  public simulateCircuit(circuit: QuantumCircuit): SimulationResult {
    if (circuit.qubitCount !== this.qubitCount) {
      throw new Error('电路量子比特数与模拟器不匹配');
    }

    this.reset();

    const sortedGates = [...circuit.gates].sort((a, b) => a.column - b.column);

    for (const gate of sortedGates) {
      this.applyGate(gate);
    }

    return {
      stateVector: this.getStateVector(),
      probabilities: this.getProbabilities(),
    };
  }

  public measure(qubitIndex: number): number {
    if (qubitIndex < 0 || qubitIndex >= this.qubitCount) {
      throw new Error(`无效的量子比特索引: ${qubitIndex}`);
    }

    const mask = 1 << (this.qubitCount - 1 - qubitIndex);
    let probability0 = 0;

    for (let i = 0; i < this.dimension; i++) {
      if ((i & mask) === 0) {
        probability0 += complexMagnitudeSq(this.stateVector[i]);
      }
    }

    const random = Math.random();
    const result = random < probability0 ? 0 : 1;

    const normalizationFactor = result === 0 ? Math.sqrt(probability0) : Math.sqrt(1 - probability0);

    for (let i = 0; i < this.dimension; i++) {
      const bitValue = (i & mask) !== 0 ? 1 : 0;
      if (bitValue !== result) {
        this.stateVector[i] = { real: 0, imag: 0 };
      } else {
        this.stateVector[i] = complexMulScalar(this.stateVector[i], 1 / normalizationFactor);
      }
    }

    return result;
  }

  public measureAll(): number {
    const probabilities = this.getProbabilities();
    const random = Math.random();
    let cumulative = 0;

    for (let i = 0; i < this.dimension; i++) {
      cumulative += probabilities[i];
      if (random <= cumulative) {
        for (let j = 0; j < this.dimension; j++) {
          this.stateVector[j] = j === i ? { real: 1, imag: 0 } : { real: 0, imag: 0 };
        }
        return i;
      }
    }

    return this.dimension - 1;
  }

  public getSingleQubitState(qubitIndex: number): { alpha: Complex; beta: Complex } {
    if (qubitIndex < 0 || qubitIndex >= this.qubitCount) {
      throw new Error(`无效的量子比特索引: ${qubitIndex}`);
    }

    const mask = 1 << (this.qubitCount - 1 - qubitIndex);
    let alpha = { real: 0, imag: 0 };
    let beta = { real: 0, imag: 0 };

    for (let i = 0; i < this.dimension; i++) {
      if ((i & mask) === 0) {
        alpha = complexAdd(alpha, this.stateVector[i]);
      } else {
        beta = complexAdd(beta, this.stateVector[i]);
      }
    }

    const norm = Math.sqrt(complexMagnitudeSq(alpha) + complexMagnitudeSq(beta));
    if (norm > 0) {
      alpha = complexMulScalar(alpha, 1 / norm);
      beta = complexMulScalar(beta, 1 / norm);
    }

    return { alpha, beta };
  }

  public applyGroverOracle(targetState: number): void {
    if (targetState < 0 || targetState >= this.dimension) {
      throw new Error(`无效的目标态: ${targetState}`);
    }
    this.stateVector[targetState] = complexMulScalar(this.stateVector[targetState], -1);
  }

  public applyGroverDiffusion(): void {
    let meanReal = 0;
    let meanImag = 0;

    for (const c of this.stateVector) {
      meanReal += c.real;
      meanImag += c.imag;
    }

    meanReal /= this.dimension;
    meanImag /= this.dimension;

    for (let i = 0; i < this.dimension; i++) {
      this.stateVector[i] = {
        real: 2 * meanReal - this.stateVector[i].real,
        imag: 2 * meanImag - this.stateVector[i].imag,
      };
    }
  }

  public runGroverSearch(targetState: number): void {
    this.reset();

    for (let i = 0; i < this.qubitCount; i++) {
      this.applySingleQubitGate('H', i);
    }

    const iterations = Math.floor(Math.PI / 4 * Math.sqrt(this.dimension));

    for (let i = 0; i < iterations; i++) {
      this.applyGroverOracle(targetState);
      this.applyGroverDiffusion();
    }
  }
}

export const simulateCircuit = (circuit: QuantumCircuit): SimulationResult => {
  const simulator = new QuantumSimulator(circuit.qubitCount);
  return simulator.simulateCircuit(circuit);
};

File: 39/wasm/quantum_simulator.cpp
#include "quantum_simulator.h"
#include <cmath>
#include <stdexcept>
#include <algorithm>

using namespace Eigen;

QuantumSimulator::QuantumSimulator(int qubitCount) : qubitCount(qubitCount) {
    if (qubitCount < 1 || qubitCount > 12) {
        throw std::invalid_argument("Qubit count must be between 1 and 12");
    }
    dimension = 1 << qubitCount;
    reset();
}

void QuantumSimulator::reset() {
    stateVector = VectorXc::Zero(dimension);
    stateVector(0) = Complex(1.0, 0.0);
}

VectorXc QuantumSimulator::getStateVector() {
    return stateVector;
}

int QuantumSimulator::getQubitCount() {
    return qubitCount;
}

std::vector<double> QuantumSimulator::getProbabilities() {
    std::vector<double> probs(dimension);
    for (int i = 0; i < dimension; i++) {
        probs[i] = std::norm(stateVector(i));
    }
    return probs;
}

MatrixXc QuantumSimulator::getGateMatrix(const std::string& gateType) {
    MatrixXc gate(2, 2);
    
    if (gateType == "H") {
        gate << Complex(1.0 / M_SQRT2, 0), Complex(1.0 / M_SQRT2, 0),
                Complex(1.0 / M_SQRT2, 0), Complex(-1.0 / M_SQRT2, 0);
    } else if (gateType == "X") {
        gate << Complex(0, 0), Complex(1, 0),
                Complex(1, 0), Complex(0, 0);
    } else if (gateType == "Y") {
        gate << Complex(0, 0), Complex(0, -1),
                Complex(0, 1), Complex(0, 0);
    } else if (gateType == "Z") {
        gate << Complex(1, 0), Complex(0, 0),
                Complex(0, 0), Complex(-1, 0);
    } else if (gateType == "T") {
        gate << Complex(1, 0), Complex(0, 0),
                Complex(0, 0), Complex(std::cos(M_PI / 4), std::sin(M_PI / 4));
    } else if (gateType == "S") {
        gate << Complex(1, 0), Complex(0, 0),
                Complex(0, 0), Complex(0, 1);
    } else if (gateType == "TDG") {
        gate << Complex(1, 0), Complex(0, 0),
                Complex(0, 0), Complex(std::cos(-M_PI / 4), std::sin(-M_PI / 4));
    } else if (gateType == "SDG") {
        gate << Complex(1, 0), Complex(0, 0),
                Complex(0, 0), Complex(0, -1);
    } else if (gateType == "I") {
        gate << Complex(1, 0), Complex(0, 0),
                Complex(0, 0), Complex(1, 0);
    } else {
        throw std::invalid_argument("Unknown gate type: " + gateType);
    }
    
    return gate;
}

MatrixXc QuantumSimulator::buildFullGateMatrix(const MatrixXc& singleQubitGate, int targetQubit) {
    MatrixXc result = MatrixXc::Identity(1, 1);
    
    for (int i = 0; i < qubitCount; i++) {
        if (i == targetQubit) {
            result = kroneckerProduct(result, singleQubitGate).eval();
        } else {
            MatrixXc identity = MatrixXc::Identity(2, 2);
            result = kroneckerProduct(result, identity).eval();
        }
    }
    
    return result;
}

void QuantumSimulator::applySingleQubitGate(const std::string& gateType, int targetQubit) {
    if (targetQubit < 0 || targetQubit >= qubitCount) {
        throw std::invalid_argument("Invalid target qubit");
    }
    
    MatrixXc gate = getGateMatrix(gateType);
    int targetMask = 1 << (qubitCount - 1 - targetQubit);
    
    VectorXc newState = VectorXc::Zero(dimension);
    
    for (int i = 0; i < dimension; i++) {
        int targetBit = (i & targetMask) ? 1 : 0;
        int otherBits = i & ~targetMask;
        
        for (int row = 0; row < 2; row++) {
            int targetIndex = otherBits | (row << (qubitCount - 1 - targetQubit));
            Complex element = gate(row, targetBit);
            newState(targetIndex) += element * stateVector(i);
        }
    }
    
    stateVector = newState;
}

void QuantumSimulator::applyCNOT(int controlQubit, int targetQubit) {
    if (controlQubit == targetQubit) {
        throw std::invalid_argument("Control and target qubits must be different");
    }
    if (controlQubit < 0 || controlQubit >= qubitCount) {
        throw std::invalid_argument("Invalid control qubit");
    }
    if (targetQubit < 0 || targetQubit >= qubitCount) {
        throw std::invalid_argument("Invalid target qubit");
    }
    
    VectorXc newState = stateVector;
    
    int controlMask = 1 << (qubitCount - 1 - controlQubit);
    int targetMask = 1 << (qubitCount - 1 - targetQubit);
    
    for (int i = 0; i < dimension; i++) {
        if (i & controlMask) {
            int flippedIndex = i ^ targetMask;
            if (i < flippedIndex) {
                std::swap(newState(i), newState(flippedIndex));
            }
        }
    }
    
    stateVector = newState;
}

int QuantumSimulator::measure(int qubitIndex) {
    if (qubitIndex < 0 || qubitIndex >= qubitCount) {
        throw std::invalid_argument("Invalid qubit index");
    }
    
    int mask = 1 << (qubitCount - 1 - qubitIndex);
    double probability0 = 0.0;
    
    for (int i = 0; i < dimension; i++) {
        if (!(i & mask)) {
            probability0 += std::norm(stateVector(i));
        }
    }
    
    double random = (double)rand() / RAND_MAX;
    int result = random < probability0 ? 0 : 1;
    
    double normalizationFactor = result == 0 ? std::sqrt(probability0) : std::sqrt(1 - probability0);
    
    for (int i = 0; i < dimension; i++) {
        int bitValue = (i & mask) ? 1 : 0;
        if (bitValue != result) {
            stateVector(i) = Complex(0, 0);
        } else {
            stateVector(i) /= normalizationFactor;
        }
    }
    
    return result;
}

int QuantumSimulator::measureAll() {
    std::vector<double> probs = getProbabilities();
    double random = (double)rand() / RAND_MAX;
    double cumulative = 0.0;
    
    for (int i = 0; i < dimension; i++) {
        cumulative += probs[i];
        if (random <= cumulative) {
            stateVector = VectorXc::Zero(dimension);
            stateVector(i) = Complex(1.0, 0.0);
            return i;
        }
    }
    
    stateVector = VectorXc::Zero(dimension);
    stateVector(dimension - 1) = Complex(1.0, 0.0);
    return dimension - 1;
}

void QuantumSimulator::applyGroverOracle(int targetState) {
    if (targetState < 0 || targetState >= dimension) {
        throw std::invalid_argument("Invalid target state");
    }
    stateVector(targetState) = -stateVector(targetState);
}

void QuantumSimulator::applyGroverDiffusion() {
    Complex mean = stateVector.mean();
    
    for (int i = 0; i < dimension; i++) {
        stateVector(i) = 2.0 * mean - stateVector(i);
    }
}

void QuantumSimulator::runGroverSearch(int targetState) {
    reset();
    
    for (int i = 0; i < qubitCount; i++) {
        applySingleQubitGate("H", i);
    }
    
    int iterations = (int)(M_PI / 4.0 * std::sqrt(dimension));
    
    for (int i = 0; i < iterations; i++) {
        applyGroverOracle(targetState);
        applyGroverDiffusion();
    }
}
